omelet-core/src/main/java/omelet/data/DataProvider.java

```diff
@@ -1,137 +1,168 @@
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
  *******************************************************************************/
 package omelet.data;
 
-import java.lang.reflect.Method;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Random;
-import java.util.Set;
-
 import omelet.common.Utils;
 import omelet.data.driverconf.IBrowserConf;
+import omelet.data.driverconf.PrepareDriverConf;
 import omelet.exception.FrameworkException;
 import omelet.testng.support.MethodContextCollection;
-
+import omelet.testng.support.RetryIAnnotationTransformer;
 import org.apache.log4j.Logger;
+import org.testng.ITestContext;
+
+import java.lang.reflect.Method;
+import java.util.*;
 
 /***
  * Data Provider class for the @Test Methods
- * 
+ *
  * @author kapilA
- * 
  */
 public class DataProvider {
+    private static final Logger LOGGER = Logger.getLogger(DataProvider.class);
+
+    public enum mapStrategy {
+        Full, Optimal
+    }
+
+    private static void updateMethodContextHolder(Method m, IBrowserConf browserConf) {
+
+        if (m.getName() != null) {
+            MethodContext context = ((MethodContext) MethodContextCollection.getMethodContext(Utils.getFullMethodName(m)));
+
+            if (browserConf != null) {
+                List<IBrowserConf> ibrowserConfList = new ArrayList<IBrowserConf>();
+                ibrowserConfList.add(browserConf);
+                context.setBrowserConf(ibrowserConfList);
+            }
+            context.prepareData();
+            RetryIAnnotationTransformer.methodContextHolder.replace(Utils.getFullMethodName(m), context);
+        }
+    }
+
+    @org.testng.annotations.DataProvider(name = "GoogleData", parallel = true)
+    public static Object[][] googleSheetDataProvider(Method m) {
+        updateMethodContextHolder(m, null);
+        String testMethodName = Utils.getFullMethodName(m);
+        return getData(testMethodName, null);
+    }
+
+    @org.testng.annotations.DataProvider(name = "XmlData", parallel = true)
+    public static Object[][] xmlDataProvider(Method m, ITestContext context) {
+        updateMethodContextHolder(m, null);
+        String methodName = Utils.getFullMethodName(m);
+
+        return getData(methodName, null);
+    }
+
+    @org.testng.annotations.DataProvider(name = "XmlSuiteData", parallel = true)
+    public static Object[][] xmlSuiteDataProvider(Method m, ITestContext context) {
+
+        String methodName = Utils.getFullMethodName(m);
+
+        Map<String, String> map = new HashMap<String, String>();
```

Comment:
This is hardcoding of the parameter , Desiredcapability is simple map which whose keys should not be restricted to certain values, do we have another way in which we can fetch the parameter keys as well from the suite.xml file?